expression/builtin_arithmetic.go

```
@@ -646,51 +647,152 @@ func (s *builtinArithmeticIntDivideDecimalSig) evalInt(row []types.Datum) (int64
+		c.setType4ModRealOrDecimal(bf.tp, tpA, tpB, true)
+		if mysql.HasUnsignedFlag(tpA.Flag) {
+			bf.tp.Flag |= mysql.UnsignedFlag
+		}
+		sig := &builtinArithmeticModDecimalSig{baseDecimalBuiltinFunc{bf}}
+		return sig.setSelf(sig), nil
+	} else {
+		bf := newBaseBuiltinFuncWithTp(args, ctx, tpInt, tpInt, tpInt)
+		if mysql.HasUnsignedFlag(tpA.Flag) {
+			bf.tp.Flag |= mysql.UnsignedFlag
+		}
+		sig := &builtinArithmeticModIntSig{baseIntBuiltinFunc{bf}}
+		return sig.setSelf(sig), nil
+	}
 }
 
-type builtinArithmeticSig struct {
-	baseBuiltinFunc
-	op opcode.Op
+type builtinArithmeticModRealSig struct {
+	baseRealBuiltinFunc
 }
 
-func (s *builtinArithmeticSig) eval(row []types.Datum) (d types.Datum, err error) {
-	args, err := s.evalArgs(row)
-	if err != nil {
-		return d, errors.Trace(err)
+func (s *builtinArithmeticModRealSig) evalReal(row []types.Datum) (float64, bool, error) {
+	sc := s.ctx.GetSessionVars().StmtCtx
+	b, isNull, err := s.args[1].EvalReal(row, sc)
+	if isNull || err != nil {
+		return 0, isNull, errors.Trace(err)
+	}
+
+	if b == 0 {
+		return 0, true, types.ErrDivByZero
 	}
+
+	a, isNull, err := s.args[0].EvalReal(row, sc)
+	if isNull || err != nil {
+		return 0, isNull, errors.Trace(err)
+	}
+
+	return math.Mod(a, b), false, nil
+}
+
+type builtinArithmeticModDecimalSig struct {
+	baseDecimalBuiltinFunc
+}
+
+func (s *builtinArithmeticModDecimalSig) evalDecimal(row []types.Datum) (*types.MyDecimal, bool, error) {
 	sc := s.ctx.GetSessionVars().StmtCtx
-	a, err := types.CoerceArithmetic(sc, args[0])
-	if err != nil {
-		return d, errors.Trace(err)
+	a, isNull, err := s.args[0].EvalDecimal(row, sc)
+	if isNull || err != nil {
+		return nil, isNull, errors.Trace(err)
+	}
+	b, isNull, err := s.args[1].EvalDecimal(row, sc)
+	if isNull || err != nil {
+		return nil, isNull, errors.Trace(err)
 	}
+	c := &types.MyDecimal{}
+	err = types.DecimalMod(a, b, c)
+	if err == types.ErrDivByZero {
+		return c, true, errors.Trace(err)
+	}
+	return c, false, nil
```

other error will be ignored ?

---

sorry，even other error should be returned. I'll fix it in next commit.